### PR TITLE
Rpm Software Filtering

### PR DESCRIPTION
**Pull request by stieg**
We have seen issues in the field where voltage echos (aka Fly-back) in the sender wires cause our unit to accidentally register an extra RPM pulse.  Here is [one such example from an Acura Integra](https://www.race-capture.com/events/24-hours-of-lemons-where-the-elite-meet-to-cheat-day-1-south-track/device/team-sheen-181?laps[]=96&laps[]=93).  As you can see, those aren't real values.  Unfortunately, this makes things unusable, so we need a fix.

This patch addresses the issue by adding in a "quiet period" into the IRQs that register the triggers.  If an IRQ occurs during this period, then we save the timer counter value locally, but ignore the pulse.  Then when the real pulse comes in, we are able to accurately calculate the period to within a much narrower window.  Multiple fly-back signals can cause a little bit of noise, but nothing so severe that it can't be overcome.

This is currently designed against 2.8.8 so we can get a solution out to the field sooner (since this is a widespread issue).  In 2.9.0 I will add support for adjusting the quiet period as part of the config.  For now I assume that the user's vehicle will not redline at 20,000 or more.

### CODE REVIEW

File: src/timer/timer.c [hunk 1]
```diff
@@ -26,14 +26,31 @@
...
+        if (MODE_LOGGING_TIMER_RPM != tc->mode)
+                return 0;
+
+        /*
+         * Software Filter Hack Represents 20K Max.  For release
+         * in 2.8.8.  Have to divide value by pulses per revolution
+         * since this must scale depending on the number of pulses
+         * per engine rotation.
+         */
+        return 3000 / tc->pulsePerRevolution;
```
Comment (brentpicasso) [inline src/timer/timer.c]: magic number for now until we have it configurable in 2.9.0?

File: stm32_base/hal/timer_stm32/timer_device_stm32.c [hunk 2]
```diff
@@ -245,107 +238,139 @@ static void init_timer_2(size_t prescaler, unsigned int slowTimerMode)
...
-    return period;
+        return channel < MK2_TIMER_CHANNELS ? g_state[channel].period : 0;
+}
+
+
+/*
+ * = = = IRQ methods below this point = = =
+ */
+
+void update_device_state(const size_t chan, const uint16_t p_ticks,
```
Comment (brentpicasso) [inline stm32_base/hal/timer_stm32/timer_device_stm32.c]: please make the p_ticks and p_h_ticks variables more verbose, or document the function. That would help

**Comment by kosenko-max** [comment]
Is this going to disallow somehow reading 0-4500Hz turbo speed sensor?

**Comment by stieg** [comment]
No.  Just change the timing mode to Hz instead of RPM.  This will bypass the filter.

**Comment by kosenko-max** [comment]
Cool. Thanks.

**Pull request merged**